### PR TITLE
feat: add `blas/base/zswap`

### LINKED ISSUE #2047: [RFC]: add `blas/base/zswap`
### Description

This RFC proposes adding the package `@stdlib/blas/base/zswap` for interchanging two complex double-precision floating-point vectors.
```c
void c_zswap( const int N, void *X, const int strideX, void *Y, const int strideY )
```

### Related Issues

Related issues: #2039

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by performant23**
If applied, this commit will add the package `blas/base/zswap`

## Description

> What is the purpose of this pull request?
Add the package `blas/base/zswap`

This pull request:
- includes the support for `C` and `Fortran` and obtains Fortran reference from [this page](https://netlib.org/lapack/explore-html-3.6.1/d2/df9/group__complex16__blas__level1_ga13a187010a0cae1fef2820072404e857.html#ga13a187010a0cae1fef2820072404e857)

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #2047 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by performant23** [comment]
Refactoring the code to use `real` instead of `realf` as reviewed in `blas/base/zcopy`, will reopen once done.

**Comment by kgryte** [comment]
@performant23 Looks like you are missing the `Makefile` for the `/src` directory.

File: lib/node_modules/@stdlib/blas/base/zswap/README.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+im = imag( z );
+// returns 0.0
+```
+
+The function has the following additional parameters:
+
+-   **offsetX**: starting index for `x`.
+-   **offsetY**: starting index for `y`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to interchange every other value in `x` starting from the second value into the last `N` elements in `y` where `x[ i ] = y[ n ]`, `x[ i+2 ] = y[ n-1 ]`, and so on.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/README.md]:
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to interchange every other value in `x` starting from the second value into the last `N` elements in `y` where `x[i] = y[n]`, `x[i+2] = y[n-1]`, and so on,
```

File: lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  ! @param {integer} iterations - number of iterations
+  ! @param {integer} len - array length
+  ! @return {double} elapsed time in seconds
+  ! ..
+  double precision function benchmark( iterations, len )
+    ! ..
+    ! External functions:
+    interface
+      subroutine zswap( N, zx, strideX, zy, strideY )
+        complex(kind=8) :: zx(:), zy(:)
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f]:
```suggestion
        complex(kind=kind(0.0d0)) :: zx(:), zy(:)
```

File: lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    ! Intrinsic functions:
+    intrinsic random_number, cpu_time, cmplx
+    ! ..
+    ! Allocate arrays:
+    allocate( x(len), y(len) )
+    ! ..
+    do i = 1, len
+      call random_number( r1 )
+      call random_number( r2 )
+      x( i ) = cmplx( (real(r1)*10000.0d0)-5000.0d0, (real(r2)*10000.0d0)-5000.0d0, kind=8 )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f]:
```suggestion
      x( i ) = cmplx( (r1*10000.0d0)-5000.0d0, (r2*10000.0d0)-5000.0d0, kind=kind(0.0d0) )
```

File: lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        call print_results( iter, elapsed )
+        ! ..
+        write (str, '(I15)') count
+        tmp = adjustl( str )
+        print '(A,A,A)', 'ok ', trim( tmp ), ' benchmark finished'
+      end do
+    end do
+    call print_summary( count, count )
+  end subroutine main
+end program bench
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f:217]:
```suggestion
end program bench

```

File: lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    ! Scalar arguments:
+    integer, intent(in) :: iterations, len
+    ! ..
+    ! Local scalars:
+    double precision :: elapsed, r1, r2
+    real :: t1, t2
+    integer :: i
+    ! ..
+    ! Local arrays:
+    complex(kind=8), allocatable :: x(:), y(:)
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/benchmark/fortran/benchmark.length.f]:
```suggestion
    complex(kind=kind(0.0d0)), allocatable :: x(:), y(:)
```

File: lib/node_modules/@stdlib/blas/base/zswap/package.json [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    "interchange",
+    "vector",
+    "typed",
+    "array",
+    "ndarray",
+    "complex",
+    "complex128",
+    "float",
+    "float64",
+    "single",
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/package.json]:
```suggestion
    "double",
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+* @param N        number of indexed elements
+* @param X        first input array
+* @param strideX  X stride length
+* @param Y        second input array
+* @param strideY  Y stride length
+*/
+void c_zswap( const int N, void *X, const int strideX, void *Y, const int strideY ) {
+	double *x = (double *)X;
+	double *y = (double *)Y;
+	double tmpR, tmpI;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c]: Prefer matching how you implemented JS with only one `tmp` variable.

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+! > Like all software, it is copyrighted. It is not trademarked, but we do ask the following:
+! >
+! > * If you modify the source for these routines we ask that you change the name of the routine and comment the changes made to the original.
+! >
+! > * We will gladly answer any questions regarding the software. If a modification is done, however, it is the responsibility of the person who modified the routine to provide support.
+!
+! @param {integer} N - number of indexed elements
+! @param {Array<complex(kind=8)>} zx - first input array
+! @param {integer} strideX - `zx` stride length
+! @param {Array<complex(kind=8)>} zy - second input array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f]:
```suggestion
! @param {Array<complex<double>>} zx - first input array
! @param {integer} strideX - `zx` stride length
! @param {Array<complex<double>>} zy - second input array
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+! @param {integer} strideY - `zy` stride length
+!<
+subroutine zswap( N, zx, strideX, zy, strideY )
+  implicit none
+  ! Scalar arguments:
+  integer :: strideX, strideY, N
+  ! Array arguments:
+  complex(kind=8), dimension(:) :: zx, zy
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f]:
```suggestion
  complex(kind=kind(0.0d0)), dimension(:) :: zx, zy
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  ! Local scalars:
+  integer :: ix, iy, i
+  complex(kind=8) :: tmp
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.f]:
```suggestion
  complex(kind=kind(0.0d0)) :: tmp
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c (hunk 7 above)
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c]:
```suggestion
	double tmp;
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+		return;
+	}
+	if ( strideX == 1 && strideY == 1 ) {
+		for ( i = 0; i < N*2; i += 2 ) {
+			tmpR = x[ i ];
+			tmpI = x[ i+1 ];
+			x[ i ] = y[ i ];
+			x[ i+1 ] = y[ i+1 ];
+			y[ i ] = tmpR;
+			y[ i+1 ] = tmpI;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c]:
```suggestion
			tmp = x[ i ];
			x[ i ] = y[ i ];
			y[ i ] = tmp;

			tmp = x[ i+1 ];
			x[ i+1 ] = y[ i+1 ];
			y[ i+1 ] = tmp;
```

File: lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	} else {
+		iy = 0;
+	}
+	for ( i = 0; i < N; i++ ) {
+		tmpR = x[ ix ];
+		tmpI = x[ ix+1 ];
+		x[ ix ] = y[ iy ];
+		x[ ix+1 ] = y[ iy+1 ];
+		y[ iy ] = tmpR;
+		y[ iy+1 ] = tmpI;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/blas/base/zswap/src/zswap.c]:
```suggestion
		tmp = x[ ix ];
		x[ ix ] = y[ iy ];
		y[ iy ] = tmp;

		tmp = x[ ix+1 ];
		x[ ix+1 ] = y[ iy+1 ];
		y[ iy+1 ] = tmp;
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @performant23!

**Comment by kgryte** [comment]
Ah, wait. You need to add the missing `Makefile` before we can merge.

**Comment by performant23** [comment]
> Ah, wait. You need to add the missing `Makefile` before we can merge.

Added, thank you so much for your review!

Edit: Ah, some tests are failing, working on the fix

Hmm, that's strange. everything seems to work locally 
image

**Comment by kgryte** [comment]
@performant23 As you are on Windows, you were testing against the C implementation which did not have any issues. However, the Fortran implementation did have bugs, and the Fortran implementation was what was being tested against in CI. The issues should now be addressed.

**Review by kgryte — APPROVED** [review]
This LGTM. Thanks, @performant23! I'll merge once CI passes.

**Pull request merged**